### DISCUSSION TITLE
Is there a way to set my custom material as default material of gltfio?

### DISCUSSION THREAD

**Discussion by SGTjeong** [google/filament#5652] [General]
I am trying to make my gltf model look like cartoon.

I tried unlit shading model, and the result color wasn't like what I expected.

![a](https://user-images.githubusercontent.com/57657823/88135156-5927f780-cc21-11ea-97fd-93f3457a8ab3.png)

The very left one is of lit shading model. The center one is of unlit shading model.

As you can see, the color of the center one is lighter, and I guess maybe sRGB color isn't being linearized well in unlit shading model.

So in fragment part, I applied **inverseTonemap(float3)**. The result is the very right one, and the color is just like what I see in Maya. 

So my question is,

**1)** In unlit shading model, why the color of model is lighter? So is it again related to sRGB EOTF? And why is applying inverseTonemap able to fix the problem?

**2)** Since gltfio uses unlit shading model as default, I should use my own custom material. If I change material with RenderableMagnager.setMaterialInstanceAt(), I need to give sampler2d Texture again. So here's the problem. If the model has a lot of meshes, materials, I need to access all of them and give texture images to each of them. 
Therefore, **I wish i could set my custom material as default material of gltfio.** What should i do?

**Comment by romainguy** [comment]
1. It is related to the sRGB OETF. You must make sure you apply the OETF before handing the colors to Filament. If you use a texture, make sure it's an sRGB texture. `inverseTonemap()` assumes the color is in linear sRGB and I won't go into details but it basically applies an sRGB OETF on the input color. Ideally you would apply the OETF yourself (or let the texture unit do it) and use `inverseTonemapSRGB()` instead (which is cheaper).

2. For that question I'll defer to @prideout but when would the default material be applied?

**Comment by SGTjeong** [comment]
I am referencing Modelviewer's loadModelGlb, loading glb model just like gltf-viewer sample.

When i have to change material of model, and setMaterialInstance(), texture information is gone, so I need new .png file for creating texture and apply to the my own material.

We are handling only glb file. because we can do things with only one glb file unlike gltf.

So I want to know if it's possible that at the very first time when assetloader and resoureloader loads glb, I could use my own material.

Or, If i could get some information about formal texture so that I won't need .png texture image even though I am using glb file.

**Comment by prideout** [comment]
C++ clients can provide custom materials to gltfio by subclassing [MaterialProvider](https://github.com/google/filament/blob/main/libs/gltfio/include/gltfio/MaterialProvider.h#L110).  Java clients do not yet have this ability, but I was thinking of adding a callback that consumes a [MaterialKey](https://github.com/google/filament/blob/main/libs/gltfio/include/gltfio/MaterialProvider.h#L39) and produces a `Material`.

**Comment by SGTjeong** [comment]
Oh yeah, it would be really cool if that function is added.
I will be looking forward to any good news, Thanks

**Comment by prideout** [comment]
@SGTjeong can you give more detail about the customization ability that you need?  Here is one proposal:

---

### Proposal 1
The callback takes four arguments: the glTF material name, the glTF node name, the glTF mesh name, and the primitive index.  It returns either a `MaterialInstance` or null.  The callback is triggered during the loading process, every time gltfio needs to assign a material to a primitive.

If the callback returns null, this means "do not customize" and gltfio will simply use an ubershader as usual.

If the callback returns a non-null material instance, gltfio will assign it to the primitive and it will never call `setParameter` on the given material instance, that will be your responsibility.  This means that it will not set textures for you, or a base color multiplier, or anything.  You can do that yourself within the callback, or wherever you want.

---

Would this work for you?

**Comment by SGTjeong** [comment]
Yes, that would totally meet my requirements.

Thanks for considering my feature request.

I will look forward to this new feature.

Thanks!

**Comment by sundaranandram** [comment]
Is there an update on this? I wanted to add a fade effect to a gltfio model depending on the z distance and this solution seems  like the most straight forward way to do it.

**Comment by sundaranandram** [comment]
Also, SceneKit went with a slightly different approach. Linking it here as a point for consideration. Although the proposed method would work just fine, one advantage with the ios approach is that, for getting effects like what i wanted above, it minimizes the code needed to create these modified shaders. Also, the current material definition format can be used with little modification to achieve the modifier. new code blocks can be introduced (much like vertex and fragment) and can be interpreted as code to be executed at certain points.

https://developer.apple.com/documentation/scenekit/scnshadermodifierentrypoint

**Comment by romainguy** [comment]
That would only help if you bring in the full version of gltfio so it could generate the shaders at runtime.

Also we currently do not want to make it possible to customize the lighting, at least not until we are sure that the internal APIs are stable.

**Comment by prideout** [comment]
@sundaranandram Also note that the current API already allows you to change the material instance on any entity -- see https://github.com/google/filament/issues/3548 for an example.

**Comment by spelman7** [comment]
Just wanted to ping this issue, since it's very related to something I'm trying to do. This quote from above summarizes my issue almost exactly:

> I am referencing Modelviewer's loadModelGlb, loading glb model just like gltf-viewer sample.
> 
> When i have to change material of model, and setMaterialInstance(), texture information is gone, so I need new .png file for creating texture and apply to the my own material.
>
> We are handling only glb file. because we can do things with only one glb file unlike gltf.
>
> So I want to know if it's possible that at the very first time when assetloader and resoureloader loads glb, I could use my own material.
>
> Or, If i could get some information about formal texture so that I won't need .png texture image even though I am using glb file.

The defaults may have changed a bit since the original post above, because in my case gltfio is loading the .glb file with a default `LIT` material, and I'd prefer an `UNLIT` material. I would be happy to construct this material myself (as mentioned in @prideout's last comment about [#3548](https://github.com/google/filament/issues/3548)), but I cannot access nor copy/re-assign the texture that loads in with the .glb.

For reference, I am using Filament for web and looking to replace my current three.js implementation. In three.js, I do this exact behavior like so:

```
if (gltfNode instanceof THREE.Mesh) {
    const materialUniforms = {
      map: { type: 't', value: (gltfNode.material as any).map },
    };

    const customMaterial = new THREE.ShaderMaterial({
      uniforms: materialUniforms,
      vertexShader: MyShader.vertexShader(),
      fragmentShader: MyShader.fragmentShader(),
    });

    gltfNode.material = customMaterial;
}
```

Having the equivalent of `materialInstance.getTexture()` would be awesome. Reading some other issues/answers, I can see that it may not be possible for various technical reasons, but it would be a huge addition to the API.

**Comment by prideout** [comment]
After #5628, you can customize the ubershaders in any way you wish, assuming you build Filament yourself.  And, later this week, we'll land another change that externalizes the pre-supplied ubershader archive.  So, even if you use a pre-built Filament, you'll still be able to provide your own ubershaders.
